source/Test.h

```diff
@@ -0,0 +1,161 @@
+/* Test.h
+Copyright (c) 2019 by Peter van der Meer
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+*/
+
+#ifndef TEST_H_
+#define TEST_H_
+
+#include "Command.h"
+#include "ConditionSet.h"
+
+#include <set>
+#include <string>
+#include <vector>
+
+class DataNode;
+class Planet;
+class PlayerInfo;
+class UI;
+class System;
+
+
+
+// Class representing a single test.
+class Test {
+public:
+	// Status indicators for the test that we selected (if any).
+	enum TestStatus {STATUS_ACTIVE, STATUS_KNOWN_FAILURE, STATUS_MISSING_FEATURE};
+	
+public:
+	// Class representing a single step in a test
+	class TestStep {
+	public:
+		// The different types of teststeps.
+		enum StepType {
+			// Step that assigns a value to a condition.
+			ASSIGN,
+			// Step that verifies if a certain condition is true
+			ASSERT,
+			// Step that breaks execution of a REPEAT loop (or stops the execution
+			// of the test itself if at toplevel) when a certain condition is
+			// true.
+			BREAK_IF,
+			// Step that adds game-data, either in the config-directories or in the game directly.
+			INJECT,
+			// Step that performs input (key, mouse, command).
+			INPUT,
+			// Invalid test-step type, should not be used in tests. Used to detect issues in test-framework.
+			INVALID,
+			// Step that performs land of the players flagship.
+			LAND,
+			// Step that launches the players flagship.
+			LAUNCH,
+			// Step to perform loading of a savegame
+			LOAD_GAME,
+			// Instructs the game to set navigation / travel plan to a target system
+			NAVIGATE,
+			// Step that contains a set of test-steps below it. Repeats the
+			// steps inside it a number of times or until a break is given within
+			// the step.
+			REPEAT,
```

Comment:
I'm considering to replace the `repeat` and `wait for` (and `break if`) steps by `label`, `goto` and `branch` so that they match the syntax of conversation nodes. Question for reviewers: What do you think, would that be usefull?

Reply:
If we plan to support scripted flight sequences / game control, it's possible some syntax should be shared between the two interfaces.

(I'm not sure if it makes the most sense for the two situations to share the same implementation / command injection method though, since during a scripted mission / behavior the player by definition should not be able to influence things, but in the case of an externally-driven test it is the "player" giving commands.)

Reply:
I agree that scripted flight sequences / game control should only have access to the in-game subset of controls. Those should not be able to access menus for things like preferences or loading savegames.

The question I was actually asking here was about sharing syntax with the _current_ conversations. Conversations use the `label`, `goto` and `branch` nodes. I wasn't planning on sharing implementation yet, but I can imagine that it would help if the syntax of the nodes in conversations and in tests matches a bit more.

Reply:
I added label and branch together with the relevant TODO items. This is now tracked in the code, so resolving this conversation.